### DISCUSSION TITLE
Can't get the correct size of the element when window resizing

### DISCUSSION THREAD

**Discussion by xland** [mikke89/RmlUi#325] [Q&A] [answered]
I don't know is it an issue.
When I resize the window,
I can't get the correct size of the element.
Not only when resizing the Window the window, but also when the window size changes are done.
```c++
auto titleLabel = document->GetElementById("titleLabel");
Rml::Vector4f rect = { titleLabel->GetAbsoluteLeft(),titleLabel->GetAbsoluteTop(),titleLabel->GetClientWidth(),titleLabel->GetClientHeight() };
Rml::Log::Message(Rml::Log::LT_INFO, "hittest %f %f %f %f", rect.x, rect.y, rect.z, rect.w);
WindowHitTestArea.push_back(rect);
```

**Comment by mikke89** [comment, marked as answer]
I think the resize event here dirties the document's layout which requires a new context update before element values are valid again. You can try to manually call `Context::Update()` or `ElementDocument::UpdateDocument()` before you query their position and size.

Read more about the validity of retrieved element values in the documentation: https://mikke89.github.io/RmlUiDoc/pages/cpp_manual/elements.html#validity-of-retrieved-values

I realized this could be better emphasized in the documentation, so I've updated it now, referencing it from the context and document pages as well.

**Reply by xland** [reply]
I've tried both APIs `Context::Update()` and `ElementDocument::UpdateDocument()`.
The problem remains.

When the app starts,it looks like this:
![image](https://user-images.githubusercontent.com/2980915/173349424-6935bdad-b30a-4ab3-9cb5-6f674f702a55.png)
When resizing the app window:
![image](https://user-images.githubusercontent.com/2980915/173349623-183e81a7-82cf-40ee-bfb9-1c08c99ea1d4.png)
You see, The three buttons on the top right of the window is not visible.
When I stop resizing the window, The three buttons is shown.
I think document is resized when I stop resizing.
But I can't get the correct element size at that time.


Size getting code is in the element's `Rml::EventId::Resize` event method.

The window is a frameless window, The title bar is rendered by rml.

**Reply by mikke89** [reply]
Ah, I see. So it is only during actual resizing there are issues here, once the mouse button is released then everything is fine again.

This is really only about submitting calls to `Context::SetDimensions()` and ensuring that rendering is performed during resize. This is more of a backend/platform thing, so you should research into how to re-draw during the resize for your platform. This can be a bit heavy performance-wise and is not always desirable, in our sample backends I think we only re-draw once the resize is complete.

**Reply by mikke89** [reply]
Your screenshots look really nice by the way :)

**Reply by xland** [reply]
OK! Thanks very much!

Code is here:
```c++
document->AddEventListener(Rml::EventId::Resize, this);
//...
void TopBar::ProcessEvent(Rml::Event& event) {
	switch (event.GetId())
	{	
	case Rml::EventId::Resize: {
		document->UpdateDocument();
		auto titleLabel = document->GetElementById("titleLabel");
		Rml::Vector4f rect = { titleLabel->GetAbsoluteLeft(),titleLabel->GetAbsoluteTop(),titleLabel->GetClientWidth(),titleLabel->GetClientHeight() };
	Rml::Log::Message(Rml::Log::LT_INFO, "hittest %f %f %f %f", rect.x, rect.y, rect.z, rect.w);
		break;
	}
	default:
		break;
	}
}
```

After I finish this project, I can share it with you if you like.

**Reply by mikke89** [reply]
Yeah, please share :)
